ICE on HRTB

The following code results in an ICE on stable, beta, and nightly
```rust
#![feature(fn_traits)]
#![feature(unboxed_closures)]

trait T0<'a, A> {
    type O;
}

struct L<T> {
    f: T,
}

impl<'a, A, T> T0<'a, A> for L<T>
where
    T: FnMut<A>,
{
    type O = T::Output;
}   

trait T1: for<'r> Ty<'r> {
    fn m<'a, B: Ty<'a>, F>(&self, f: F) -> ()
    where
        F: for<'r> T0<'r, (<Self as Ty<'r>>::V,), O = <B as Ty<'r>>::V>,
    {
        unimplemented!();
    }
}

trait Ty<'a> {
    type V;
}

fn main() {
    let v = ().m(L{ f : |x|{drop(x)} } );
}

impl<'a> Ty<'a> for () {
    type V = &'a u8;
}

impl T1 for () {}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=39c45aaf72f37dcb1052e4537b96ba36))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0554]: #![feature] may not be used on the stable release channel
 --> src/main.rs:1:1
  |
1 | #![feature(fn_traits)]
  | ^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: #![feature] may not be used on the stable release channel
 --> src/main.rs:2:1
  |
2 | #![feature(unboxed_closures)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0271]: type mismatch resolving `for<'r> <L<[closure@src/main.rs:33:25: 33:37]> as T0<'r, (<() as Ty<'r>>::V,)>>::O == <_ as Ty<'r>>::V`
  --> src/main.rs:33:16
   |
33 |     let v = ().m(L{ f : |x|{drop(x)} } );
   |                ^ expected (), found associated type
   |
   = note: expected type `()`
              found type `<_ as Ty<'_>>::V`

error: internal compiler error: src/librustc/infer/lexical_region_resolve/mod.rs:632: collect_error_for_expanding_node() could not find error for var '_#5r in universe U8, lower_bounds=[
    RegionAndOrigin(RePlaceholder(Placeholder { universe: U7, name: BrNamed(crate0:DefIndex(1:20), 'r) }),Subtype(TypeTrace(ObligationCause { span: src/main.rs:33:16: 33:17, body_id: HirId { owner: DefIndex(0:12), local_id: 22 }, code: ImplDerivedObligation(DerivedObligationCause { parent_trait_ref: Binder(<L<[closure@src/main.rs:33:25: 33:37]> as T0<'r, (<() as Ty<'r>>::V,)>>), parent_code: MiscObligation }) }))),
    RegionAndOrigin(RePlaceholder(Placeholder { universe: U7, name: BrNamed(crate0:DefIndex(1:20), 'r) }),Subtype(TypeTrace(ObligationCause { span: src/main.rs:33:16: 33:17, body_id: HirId { owner: DefIndex(0:12), local_id: 22 }, code: ImplDerivedObligation(DerivedObligationCause { parent_trait_ref: Binder(<L<[closure@src/main.rs:33:25: 33:37]> as T0<'r, (<() as Ty<'r>>::V,)>>), parent_code: MiscObligation }) }))),
], upper_bounds=[
    RegionAndOrigin(RePlaceholder(Placeholder { universe: U7, name: BrNamed(crate0:DefIndex(1:20), 'r) }),Subtype(TypeTrace(ObligationCause { span: src/main.rs:33:16: 33:17, body_id: HirId { owner: DefIndex(0:12), local_id: 22 }, code: ImplDerivedObligation(DerivedObligationCause { parent_trait_ref: Binder(<L<[closure@src/main.rs:33:25: 33:37]> as T0<'r, (<() as Ty<'r>>::V,)>>), parent_code: MiscObligation }) }))),
    RegionAndOrigin(RePlaceholder(Placeholder { universe: U7, name: BrNamed(crate0:DefIndex(1:20), 'r) }),Subtype(TypeTrace(ObligationCause { span: src/main.rs:33:16: 33:17, body_id: HirId { owner: DefIndex(0:12), local_id: 22 }, code: ImplDerivedObligation(DerivedObligationCause { parent_trait_ref: Binder(<L<[closure@src/main.rs:33:25: 33:37]> as T0<'r, (<() as Ty<'r>>::V,)>>), parent_code: MiscObligation }) }))),
]
  --> src/main.rs:33:16
   |
33 |     let v = ().m(L{ f : |x|{drop(x)} } );
   |                ^

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:570:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to 4 previous errors

Some errors occurred: E0271, E0554.
For more information about an error, try `rustc --explain E0271`.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.35.0 (3c235d560 2019-05-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

--- a/src/librustc/infer/lexical_region_resolve/mod.rs
+++ b/src/librustc/infer/lexical_region_resolve/mod.rs
@@ -764,16 +764,17 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
             }
         }
 
-        span_bug!(
+        // Errors in earlier passes can yield error variables without
+        // resolution errors here; delay ICE in favor of those errors.
+        self.tcx().sess.delay_span_bug(
             self.var_infos[node_idx].origin.span(),
-            "collect_error_for_expanding_node() could not find \
-             error for var {:?} in universe {:?}, lower_bounds={:#?}, \
-             upper_bounds={:#?}",
-            node_idx,
-            node_universe,
-            lower_bounds,
-            upper_bounds
-        );
+            &format!("collect_error_for_expanding_node() could not find \
+                      error for var {:?} in universe {:?}, lower_bounds={:#?}, \
+                      upper_bounds={:#?}",
+                     node_idx,
+                     node_universe,
+                     lower_bounds,
+                     upper_bounds));
     }
 
     fn collect_concrete_regions(

--- a/src/test/ui/hrtb/issue-62203-hrtb-ice.rs
+++ b/src/test/ui/hrtb/issue-62203-hrtb-ice.rs
@@ -1,0 +1,50 @@
+trait T0<'a, A> {
+    type O;
+}
+
+struct L<T> {
+    f: T,
+}
+
+// explicitly named variants of what one would normally denote by the
+// unit type `()`. Why do this? So that we can differentiate them in
+// the diagnostic output.
+struct Unit1;
+struct Unit2;
+struct Unit3;
+struct Unit4;
+
+impl<'a, A, T> T0<'a, A> for L<T>
+where
+    T: FnMut(A) -> Unit3,
+{
+    type O = T::Output;
+}
+
+trait T1: for<'r> Ty<'r> {
+    fn m<'a, B: Ty<'a>, F>(&self, f: F) -> Unit1
+    where
+        F: for<'r> T0<'r, (<Self as Ty<'r>>::V,), O = <B as Ty<'r>>::V>,
+    {
+        unimplemented!();
+    }
+}
+
+trait Ty<'a> {
+    type V;
+}
+
+fn main() {
+    let v = Unit2.m(
+        //~^ ERROR type mismatch
+        //~| ERROR type mismatch
+        L {
+            f : |x| { drop(x); Unit4 }
+        });
+}
+
+impl<'a> Ty<'a> for Unit2 {
+    type V = &'a u8;
+}
+
+impl T1 for Unit2 {}

--- a/src/test/ui/hrtb/issue-62203-hrtb-ice.stderr
+++ b/src/test/ui/hrtb/issue-62203-hrtb-ice.stderr
@@ -1,0 +1,22 @@
+error[E0271]: type mismatch resolving `for<'r> <L<[closure@$DIR/issue-62203-hrtb-ice.rs:42:17: 42:39]> as T0<'r, (<Unit2 as Ty<'r>>::V,)>>::O == <_ as Ty<'r>>::V`
+  --> $DIR/issue-62203-hrtb-ice.rs:38:19
+   |
+LL |     let v = Unit2.m(
+   |                   ^ expected struct `Unit4`, found associated type
+   |
+   = note: expected type `Unit4`
+              found type `<_ as Ty<'_>>::V`
+
+error[E0271]: type mismatch resolving `<[closure@$DIR/issue-62203-hrtb-ice.rs:42:17: 42:39] as std::ops::FnOnce<((&u8,),)>>::Output == Unit3`
+  --> $DIR/issue-62203-hrtb-ice.rs:38:19
+   |
+LL |     let v = Unit2.m(
+   |                   ^ expected struct `Unit4`, found struct `Unit3`
+   |
+   = note: expected type `Unit4`
+              found type `Unit3`
+   = note: required because of the requirements on the impl of `for<'r> T0<'r, (<Unit2 as Ty<'r>>::V,)>` for `L<[closure@$DIR/issue-62203-hrtb-ice.rs:42:17: 42:39]>`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0271`.

